RESTDataSource docs updates (#7255)
apollographql/apollo-server: docs/source/data/fetching-rest.mdx

> > See the [`@apollo/datasource-rest` page](https://github.com/apollographql/datasource-rest) for the full details of the `RESTDataSource` API.

Reviewer: ```suggestion
> See the [`@apollo/datasource-rest` README](https://github.com/apollographql/datasource-rest) for the full details of the `RESTDataSource` API.
```
perhaps?

</MultiCodeBlock>

> You can extend the `RESTDataSource` class to implement whatever data-fetching methods your resolvers need. These methods should use the built-in convenience methods (e.g., `get` and `post`) to perform HTTP requests, helping you add query parameters, parse and cache JSON results, dedupe requests, and handle errors. More complex use cases can use the `fetch` method directly. The `fetch` method returns both the parsed body as well as the response object which affords more flexibility for use cases like reading response headers.

Reviewer 2: ```suggestion
You can extend the `RESTDataSource` class to implement whatever data-fetching methods your resolvers need. These methods should use the built-in convenience methods (e.g., `get` and `post`) to perform HTTP requests, helping you add query parameters, parse and cache JSON results, dedupe requests, and handle errors. More complex use cases can use the `fetch` method directly. The `fetch` method returns both the parsed body and the response object, which provides more flexibility for use cases like reading response headers.
```

The `RESTDataSource` class provides its subclasses with _two_ layers of caching:

> - The first layer deduplicates _concurrent_ outgoing `GET` (and `HEAD`) requests by default. Deduplication is keyed on the request's method and URL. This behavior can be configured by overriding the `requestDeduplicationPolicyFor` method. For more details on how to configure this please consult [the README](https://github.com/apollographql/datasource-rest#requestdeduplicationpolicyfor).

Reviewer 2: ```suggestion
- The first layer deduplicates _concurrent_ outgoing `GET` (and `HEAD`) requests by default. Deduplication is keyed on the request's method and URL. You can configure this behavior by overriding the `requestDeduplicationPolicyFor` method. For more details, see [the README](https://github.com/apollographql/datasource-rest#requestdeduplicationpolicyfor).
```

> > Note: in versions prior to v5 of `RESTDataSource`, _all_ outgoing `GET` requests were deduplicated. This behavior can be restored using the `deduplicate-until-invalidated` policy (explained more in-depth in [the README](https://github.com/apollographql/datasource-rest#requestdeduplicationpolicyfor)).

Reviewer 2: ```suggestion
> **Note:** In versions of `RESTDataSource` prior to v5, _all_ outgoing `GET` requests are deduplicated. You can achieve this same behavior with the `deduplicate-until-invalidated` policy (explained further in [the README](https://github.com/apollographql/datasource-rest#requestdeduplicationpolicyfor)).
```

- The second layer caches the results from **HTTP responses that specify HTTP caching headers**.

These caching layers effectively make the `RESTDataSource` class a Node HTTP client that offers browser-style caching. Below, we'll dive into each layer of caching and the advantage that layer provides.

### `GET` (and `HEAD`) requests and responses

> Every time you create a new instance of a `RESTDataSource` subclass, under the hood, it makes an internal cache. By default, `RESTDataSource` automatically deduplicates _concurrent_ `GET` (and `HEAD`) requests (keyed by their method and URLs) alongside their results in this internal cache. This is referred to as request deduplication. The default behavior can be configured by overriding the [`requestDeduplicationPolicyFor`](https://github.com/apollographql/datasource-rest/blob/main/README.md#requestdeduplicationpolicyfor) method on the class.

Reviewer 2: ```suggestion
Every time you instantiate a `RESTDataSource` subclass, under the hood that instance creates an internal cache. By default, `RESTDataSource` automatically deduplicates _concurrent_ `GET` (and `HEAD`) requests (keyed by their method and URLs) alongside their results in this internal cache. This behavior is called request deduplication. You can configure this default behavior by overriding the [`requestDeduplicationPolicyFor`](https://github.com/apollographql/datasource-rest/blob/main/README.md#requestdeduplicationpolicyfor) method on the class.
```

> This internal caching mechanism is why we create a new `RESTDataSource` instance for _every_ request. Otherwise, responses would be cached across requests even if they specify they shouldn't be!

> You can change how `GET` (and `HEAD`) requests are stored in `RESTDataSource`'s deduplication cache by overwriting the [`cacheKeyFor` method](https://github.com/apollographql/datasource-rest#cachekeyfor). By default, the cache key is the method and URL of the request.

Reviewer 2: ```suggestion
You can change how `GET` (and `HEAD`) requests are stored in `RESTDataSource`'s deduplication cache by overwriting the [`cacheKeyFor` method](https://github.com/apollographql/datasource-rest#cachekeyfor). By default, a request's cache key is the combination of its HTTP method and URL.
```

> If you'd like to restore the deduplication policy from before `RESTDataSource` v5, you can configure `requestDeduplicationPolicyFor` like so:

Reviewer 2: ```suggestion
To restore the deduplication policy from before `RESTDataSource` v5, you can configure `requestDeduplicationPolicyFor` like so:
```
